// @cloud-carbon-footprint/core/dist/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.FootprintEstimatesDataBuilder = exports.Region = exports.BillingDataRow = exports.AccumulateKilowattHoursBy = exports.estimateKwh = exports.estimateCo2 = exports.getAverage = exports.getWattsByAverageOrMedian = exports.aggregateEstimatesByDay = exports.accumulateKilowattHours = exports.appendOrAccumulateEstimatesByDay = exports.US_NERC_REGIONS_EMISSIONS_FACTORS = exports.CloudProviderAccount = void 0;
__exportStar(require("./memory"), exports);
__exportStar(require("./networking"), exports);
__exportStar(require("./storage"), exports);
__exportStar(require("./compute"), exports);
__exportStar(require("./cost"), exports);
__exportStar(require("./unknown"), exports);
__exportStar(require("./embodiedEmissions"), exports);
var CloudProviderAccount_1 = require("./CloudProviderAccount");
Object.defineProperty(exports, "CloudProviderAccount", { enumerable: true, get: function () { return __importDefault(CloudProviderAccount_1).default; } });
var FootprintEstimationConstants_1 = require("./FootprintEstimationConstants");
Object.defineProperty(exports, "US_NERC_REGIONS_EMISSIONS_FACTORS", { enumerable: true, get: function () { return FootprintEstimationConstants_1.US_NERC_REGIONS_EMISSIONS_FACTORS; } });
var FootprintEstimate_1 = require("./FootprintEstimate");
Object.defineProperty(exports, "appendOrAccumulateEstimatesByDay", { enumerable: true, get: function () { return FootprintEstimate_1.appendOrAccumulateEstimatesByDay; } });
Object.defineProperty(exports, "accumulateKilowattHours", { enumerable: true, get: function () { return FootprintEstimate_1.accumulateKilowattHours; } });
Object.defineProperty(exports, "aggregateEstimatesByDay", { enumerable: true, get: function () { return FootprintEstimate_1.aggregateEstimatesByDay; } });
Object.defineProperty(exports, "getWattsByAverageOrMedian", { enumerable: true, get: function () { return FootprintEstimate_1.getWattsByAverageOrMedian; } });
Object.defineProperty(exports, "getAverage", { enumerable: true, get: function () { return FootprintEstimate_1.getAverage; } });
Object.defineProperty(exports, "estimateCo2", { enumerable: true, get: function () { return FootprintEstimate_1.estimateCo2; } });
Object.defineProperty(exports, "estimateKwh", { enumerable: true, get: function () { return FootprintEstimate_1.estimateKwh; } });
Object.defineProperty(exports, "AccumulateKilowattHoursBy", { enumerable: true, get: function () { return FootprintEstimate_1.AccumulateKilowattHoursBy; } });
var BillingDataRow_1 = require("./BillingDataRow");
Object.defineProperty(exports, "BillingDataRow", { enumerable: true, get: function () { return __importDefault(BillingDataRow_1).default; } });
var Region_1 = require("./Region");
Object.defineProperty(exports, "Region", { enumerable: true, get: function () { return __importDefault(Region_1).default; } });
var FootprintEstimatesDataBuilder_1 = require("./FootprintEstimatesDataBuilder");
Object.defineProperty(exports, "FootprintEstimatesDataBuilder", { enumerable: true, get: function () { return __importDefault(FootprintEstimatesDataBuilder_1).default; } });
