// @grnsft/if-unofficial-plugins/build/lib/boavizta/boavizta-api.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.BoaviztaAPI = void 0;
const axios_1 = require("axios");
const errors_1 = require("../../util/errors");
const helpers_1 = require("../../util/helpers");
const { APIRequestError } = errors_1.ERRORS;
const BoaviztaAPI = () => {
    const BASE_URL = 'https://api.boavizta.org/v1';
    const errorBuilder = (0, helpers_1.buildErrorMessage)(exports.BoaviztaAPI.name);
    /**
     * Fetches CPU output data from Boavizta API for a specific component type.
     */
    const fetchCpuOutputData = async (data, componentType, verbose) => {
        const dataCast = {
            core_units: data['cpu/number-cores'],
            name: data['cpu/name'],
            tdp: data['cpu/thermal-design-power'],
        };
        const response = await axios_1.default.post(`${BASE_URL}/component/${componentType}?verbose=${verbose}&duration=${data['usage']['hours_use_time']}`, dataCast);
        return response.data;
    };
    /**
     * Fetches cloud instance data from Boavizta API.
     */
    const fetchCloudInstanceData = async (data, verbose) => {
        const updatedDataCast = replaceHyphensWithUnderscores(data);
        const response = await axios_1.default
            .post(`${BASE_URL}/cloud/instance?verbose=${verbose}&duration=${updatedDataCast['usage']['hours_use_time']}`, {
            provider: data.provider,
            instance_type: data['instance-type'],
            usage: data.usage,
        })
            .catch(error => {
            throw new APIRequestError(errorBuilder({
                message: `Error fetching data from Boavizta API. ${JSON.stringify((error.response &&
                    error.response.data &&
                    error.response.data.detail) ||
                    error)}`,
            }));
        });
        return response?.data;
    };
    /**
     * Gets the list of supported cloud instances for a given provider.
     */
    const getSupportedInstancesList = async (provider) => {
        const instances = await axios_1.default.get(`${BASE_URL}/cloud/instance/all_instances?provider=${provider}`);
        return instances.data;
    };
    /**
     * Gets the list of supported cloud providers.
     */
    const getSupportedProvidersList = async () => {
        const providers = await axios_1.default.get(`${BASE_URL}/cloud/instance/all_providers`);
        return providers.data;
    };
    /**
     * Gets the list of supported locations by the model.
     */
    const getSupportedLocations = async () => {
        const countries = await axios_1.default.get(`${BASE_URL}/utils/country_code`);
        return Object.values(countries.data);
    };
    /**
     * Replaces hyphens with underscores in keys of a key-value pair object.
     */
    const replaceHyphensWithUnderscores = (data) => {
        return Object.fromEntries(Object.entries(data).map(([key, value]) => [
            key.replace(/-/g, '_'),
            value,
        ]));
    };
    return {
        getSupportedLocations,
        getSupportedProvidersList,
        getSupportedInstancesList,
        fetchCloudInstanceData,
        fetchCpuOutputData,
    };
};
exports.BoaviztaAPI = BoaviztaAPI;
