// @cloud-carbon-footprint/core/dist/compute/ComputeProcessorTypes.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.skyLakeBroadwellHaswellSandyBridgeIvyBridge = exports.skyLakeBroadwellHaswellAMDRome = exports.cascadeLakeHaswell = exports.cascadeLakeSkylakeBroadwell = exports.cascadeLakeSkylakeBroadwellHaswell = exports.broadwelCascadeLake = exports.broadwellSkylake = exports.broadwellHaswell = exports.cascadeLakeSkylake = exports.COMPUTE_PROCESSOR_TYPES = void 0;
var COMPUTE_PROCESSOR_TYPES;
(function (COMPUTE_PROCESSOR_TYPES) {
    COMPUTE_PROCESSOR_TYPES["CASCADE_LAKE"] = "Cascade Lake";
    COMPUTE_PROCESSOR_TYPES["SKYLAKE"] = "Skylake";
    COMPUTE_PROCESSOR_TYPES["HASWELL"] = "Haswell";
    COMPUTE_PROCESSOR_TYPES["BROADWELL"] = "Broadwell";
    COMPUTE_PROCESSOR_TYPES["COFFEE_LAKE"] = "Coffee Lake";
    COMPUTE_PROCESSOR_TYPES["SANDY_BRIDGE"] = "Sandy Bridge";
    COMPUTE_PROCESSOR_TYPES["IVY_BRIDGE"] = "Ivy Bridge";
    COMPUTE_PROCESSOR_TYPES["ICELAKE"] = "Ice Lake";
    COMPUTE_PROCESSOR_TYPES["AMD_EPYC_1ST_GEN"] = "AMD EPYC 1st Gen";
    COMPUTE_PROCESSOR_TYPES["AMD_EPYC_2ND_GEN"] = "AMD EPYC 2nd Gen";
    COMPUTE_PROCESSOR_TYPES["AMD_EPYC_3RD_GEN"] = "AMD EPYC 3rd Gen";
    COMPUTE_PROCESSOR_TYPES["AWS_GRAVITON"] = "AWS Graviton";
    COMPUTE_PROCESSOR_TYPES["AWS_GRAVITON_2"] = "AWS Graviton2";
    COMPUTE_PROCESSOR_TYPES["NVIDIA_K520"] = "Nvidia K520";
    COMPUTE_PROCESSOR_TYPES["NVIDIA_A10G"] = "Nvidia A10G";
    COMPUTE_PROCESSOR_TYPES["NVIDIA_T4"] = "Nvidia T4";
    COMPUTE_PROCESSOR_TYPES["NVIDIA_TESLA_M60"] = "Nvidia Tesla M60";
    COMPUTE_PROCESSOR_TYPES["NVIDIA_TESLA_K80"] = "Nvidia Tesla K80";
    COMPUTE_PROCESSOR_TYPES["NVIDIA_TESLA_V100"] = "Nvidia Tesla V100";
    COMPUTE_PROCESSOR_TYPES["NVIDIA_TESLA_A100"] = "Nvidia Tesla A100";
    COMPUTE_PROCESSOR_TYPES["NVIDIA_TESLA_P4"] = "Nvidia Tesla P4";
    COMPUTE_PROCESSOR_TYPES["NVIDIA_TESLA_P40"] = "Nvidia Tesla P40";
    COMPUTE_PROCESSOR_TYPES["NVIDIA_TESLA_P100"] = "Nvidia Tesla P100";
    COMPUTE_PROCESSOR_TYPES["AMD_RADEON_PRO_V520"] = "AMD Radeon Pro V520";
    COMPUTE_PROCESSOR_TYPES["XILINX_ALVEO_U250"] = "Xilinx Alveo U250";
    COMPUTE_PROCESSOR_TYPES["UNKNOWN"] = "Unknown";
})(COMPUTE_PROCESSOR_TYPES = exports.COMPUTE_PROCESSOR_TYPES || (exports.COMPUTE_PROCESSOR_TYPES = {}));
exports.cascadeLakeSkylake = [
    COMPUTE_PROCESSOR_TYPES.CASCADE_LAKE,
    COMPUTE_PROCESSOR_TYPES.SKYLAKE,
];
exports.broadwellHaswell = [
    COMPUTE_PROCESSOR_TYPES.HASWELL,
    COMPUTE_PROCESSOR_TYPES.BROADWELL,
];
exports.broadwellSkylake = [
    COMPUTE_PROCESSOR_TYPES.BROADWELL,
    COMPUTE_PROCESSOR_TYPES.SKYLAKE,
];
exports.broadwelCascadeLake = [
    COMPUTE_PROCESSOR_TYPES.BROADWELL,
    COMPUTE_PROCESSOR_TYPES.CASCADE_LAKE,
];
exports.cascadeLakeSkylakeBroadwellHaswell = [
    COMPUTE_PROCESSOR_TYPES.CASCADE_LAKE,
    COMPUTE_PROCESSOR_TYPES.SKYLAKE,
    COMPUTE_PROCESSOR_TYPES.BROADWELL,
    COMPUTE_PROCESSOR_TYPES.HASWELL,
];
exports.cascadeLakeSkylakeBroadwell = [
    COMPUTE_PROCESSOR_TYPES.CASCADE_LAKE,
    COMPUTE_PROCESSOR_TYPES.SKYLAKE,
    COMPUTE_PROCESSOR_TYPES.BROADWELL,
];
exports.cascadeLakeHaswell = [
    COMPUTE_PROCESSOR_TYPES.HASWELL,
    COMPUTE_PROCESSOR_TYPES.CASCADE_LAKE,
];
exports.skyLakeBroadwellHaswellAMDRome = [
    COMPUTE_PROCESSOR_TYPES.SKYLAKE,
    COMPUTE_PROCESSOR_TYPES.BROADWELL,
    COMPUTE_PROCESSOR_TYPES.HASWELL,
    COMPUTE_PROCESSOR_TYPES.AMD_EPYC_2ND_GEN,
];
exports.skyLakeBroadwellHaswellSandyBridgeIvyBridge = [
    COMPUTE_PROCESSOR_TYPES.SKYLAKE,
    COMPUTE_PROCESSOR_TYPES.BROADWELL,
    COMPUTE_PROCESSOR_TYPES.HASWELL,
    COMPUTE_PROCESSOR_TYPES.SANDY_BRIDGE,
    COMPUTE_PROCESSOR_TYPES.IVY_BRIDGE,
];
