// @cloud-carbon-footprint/common/dist/Errors.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.PartialDataError = exports.RecommendationsRequestValidationError = exports.EstimationRequestValidationError = exports.ErrorTypes = void 0;
var ErrorTypes;
(function (ErrorTypes) {
    ErrorTypes["PARTIAL_DATA_ERROR"] = "PartialDataError";
    ErrorTypes["ESTIMATION_REQUEST_VALIDATION_ERROR"] = "EstimationRequestValidationError";
    ErrorTypes["RECOMMENDATIONS_REQUEST_VALIDATION_ERROR"] = "RecommendationsRequestValidationError";
})(ErrorTypes = exports.ErrorTypes || (exports.ErrorTypes = {}));
class EstimationRequestValidationError extends Error {
    constructor(message) {
        super(message);
        this.name = ErrorTypes.ESTIMATION_REQUEST_VALIDATION_ERROR;
        Object.setPrototypeOf(this, EstimationRequestValidationError.prototype);
    }
}
exports.EstimationRequestValidationError = EstimationRequestValidationError;
class RecommendationsRequestValidationError extends Error {
    constructor(message) {
        super(message);
        this.name = ErrorTypes.RECOMMENDATIONS_REQUEST_VALIDATION_ERROR;
        Object.setPrototypeOf(this, RecommendationsRequestValidationError.prototype);
    }
}
exports.RecommendationsRequestValidationError = RecommendationsRequestValidationError;
class PartialDataError extends Error {
    constructor(message) {
        super(message);
        this.name = ErrorTypes.PARTIAL_DATA_ERROR;
        Object.setPrototypeOf(this, PartialDataError.prototype);
    }
}
exports.PartialDataError = PartialDataError;
