// @cloud-carbon-footprint/common/dist/EstimationResult.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.reduceByTimestamp = void 0;
const ramda_1 = require("ramda");
const helpers_1 = require("./helpers");
const reduceByTimestamp = (estimationResults) => {
    const accumulatingFn = (acc, value) => {
        acc.timestamp = acc.timestamp || new Date(value.timestamp);
        acc.serviceEstimates = acc.serviceEstimates.concat(value.serviceEstimates);
        acc.groupBy = acc.groupBy || value.groupBy;
        acc.periodStartDate = acc.timestamp || new Date(value.timestamp);
        acc.periodEndDate = (0, helpers_1.getPeriodEndDate)(acc.periodStartDate, acc.groupBy);
        return acc;
    };
    const getTimeOfEstimate = (estimationResult) => estimationResult.timestamp.toISOString();
    const result = (0, ramda_1.reduceBy)(accumulatingFn, {
        timestamp: undefined,
        serviceEstimates: [],
        periodStartDate: undefined,
        periodEndDate: undefined,
        groupBy: undefined,
    }, getTimeOfEstimate, estimationResults);
    return Object.values(result);
};
exports.reduceByTimestamp = reduceByTimestamp;
