// @cloud-carbon-footprint/common/dist/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.AWS_DEFAULT_RECOMMENDATIONS_SERVICE = exports.AWS_RECOMMENDATIONS_SERVICES = exports.AWS_DEFAULT_RECOMMENDATION_TARGET = exports.AWS_RECOMMENDATIONS_TARGETS = exports.reduceByTimestamp = exports.RecommendationsRequestValidationError = exports.EstimationRequestValidationError = exports.PartialDataError = exports.GroupBy = exports.Config = exports.setConfig = exports.configLoader = exports.Logger = void 0;
var Logger_1 = require("./Logger");
Object.defineProperty(exports, "Logger", { enumerable: true, get: function () { return __importDefault(Logger_1).default; } });
var ConfigLoader_1 = require("./ConfigLoader");
Object.defineProperty(exports, "configLoader", { enumerable: true, get: function () { return __importDefault(ConfigLoader_1).default; } });
Object.defineProperty(exports, "setConfig", { enumerable: true, get: function () { return ConfigLoader_1.setConfig; } });
var Config_1 = require("./Config");
Object.defineProperty(exports, "Config", { enumerable: true, get: function () { return __importDefault(Config_1).default; } });
var Config_2 = require("./Config");
Object.defineProperty(exports, "GroupBy", { enumerable: true, get: function () { return Config_2.GroupBy; } });
var Errors_1 = require("./Errors");
Object.defineProperty(exports, "PartialDataError", { enumerable: true, get: function () { return Errors_1.PartialDataError; } });
Object.defineProperty(exports, "EstimationRequestValidationError", { enumerable: true, get: function () { return Errors_1.EstimationRequestValidationError; } });
Object.defineProperty(exports, "RecommendationsRequestValidationError", { enumerable: true, get: function () { return Errors_1.RecommendationsRequestValidationError; } });
var EstimationResult_1 = require("./EstimationResult");
Object.defineProperty(exports, "reduceByTimestamp", { enumerable: true, get: function () { return EstimationResult_1.reduceByTimestamp; } });
var RecommendationTarget_1 = require("./RecommendationTarget");
Object.defineProperty(exports, "AWS_RECOMMENDATIONS_TARGETS", { enumerable: true, get: function () { return RecommendationTarget_1.AWS_RECOMMENDATIONS_TARGETS; } });
Object.defineProperty(exports, "AWS_DEFAULT_RECOMMENDATION_TARGET", { enumerable: true, get: function () { return RecommendationTarget_1.AWS_DEFAULT_RECOMMENDATION_TARGET; } });
var RecommendationsService_1 = require("./RecommendationsService");
Object.defineProperty(exports, "AWS_RECOMMENDATIONS_SERVICES", { enumerable: true, get: function () { return RecommendationsService_1.AWS_RECOMMENDATIONS_SERVICES; } });
Object.defineProperty(exports, "AWS_DEFAULT_RECOMMENDATIONS_SERVICE", { enumerable: true, get: function () { return RecommendationsService_1.AWS_DEFAULT_RECOMMENDATIONS_SERVICE; } });
__exportStar(require("./helpers"), exports);
__exportStar(require("./EmissionsFactors"), exports);
