// @grnsft/if-unofficial-plugins/build/lib/azure-importer/azure-api.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.AzureAPI = void 0;
const arm_monitor_1 = require("@azure/arm-monitor");
const identity_1 = require("@azure/identity");
const arm_compute_1 = require("@azure/arm-compute");
class AzureAPI {
    constructor() {
        this.credential = new identity_1.DefaultAzureCredential();
    }
    /**
     * Fetches metrics for a specific virtual machine.
     */
    async getMetricsTimeseries(params, metricName) {
        const { subscriptionId, resourceGroupName, timespan, interval, aggregation, vmName, } = params;
        this.setClient(subscriptionId);
        const monitorClient = new arm_monitor_1.MonitorClient(this.credential, subscriptionId);
        const response = await monitorClient.metrics.list(`subscriptions/${subscriptionId}/resourceGroups/${resourceGroupName}/providers/Microsoft.Compute/virtualMachines/${vmName}`, {
            metricnames: metricName,
            timespan,
            interval,
            aggregation,
        });
        return (response.value[0].timeseries || [])
            .map(series => series.data || [])
            .flat();
    }
    /**
     * Sets up the ComputeManagementClient with the specified subscription ID.
     */
    setClient(subscriptionId) {
        this.client = new arm_compute_1.ComputeManagementClient(this.credential, subscriptionId);
    }
    /**
     * Fetches a list of virtual machines within a specific resource group.
     */
    async getVMDataByResourceGroupName(resourceGroupName) {
        const vmData = [];
        const vmList = this.client.virtualMachines.list(resourceGroupName);
        for await (const item of vmList) {
            vmData.push(item);
        }
        return vmData;
    }
    /**
     * Fetches a list of available resource Skus.
     */
    async getResourceSkus() {
        const iterator = this.client.resourceSkus.list();
        const resourceSkus = [];
        for await (const resourceSku of iterator) {
            resourceSkus.push(resourceSku);
        }
        return resourceSkus;
    }
}
exports.AzureAPI = AzureAPI;
