// @cloud-carbon-footprint/aws/dist/lib/index.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.ComputeOptimizerRecommendations = exports.RightsizingRecommendations = exports.AWS_REGIONS = exports.CostAndUsageReports = exports.ServiceWrapper = exports.Lambda = exports.RDSStorage = exports.RDSComputeService = exports.RDS = exports.ElastiCache = exports.EC2 = exports.S3 = exports.EBS = void 0;
var EBS_1 = require("./EBS");
Object.defineProperty(exports, "EBS", { enumerable: true, get: function () { return __importDefault(EBS_1).default; } });
var S3_1 = require("./S3");
Object.defineProperty(exports, "S3", { enumerable: true, get: function () { return __importDefault(S3_1).default; } });
var EC2_1 = require("./EC2");
Object.defineProperty(exports, "EC2", { enumerable: true, get: function () { return __importDefault(EC2_1).default; } });
var ElastiCache_1 = require("./ElastiCache");
Object.defineProperty(exports, "ElastiCache", { enumerable: true, get: function () { return __importDefault(ElastiCache_1).default; } });
var RDS_1 = require("./RDS");
Object.defineProperty(exports, "RDS", { enumerable: true, get: function () { return __importDefault(RDS_1).default; } });
var RDSCompute_1 = require("./RDSCompute");
Object.defineProperty(exports, "RDSComputeService", { enumerable: true, get: function () { return __importDefault(RDSCompute_1).default; } });
var RDSStorage_1 = require("./RDSStorage");
Object.defineProperty(exports, "RDSStorage", { enumerable: true, get: function () { return __importDefault(RDSStorage_1).default; } });
var Lambda_1 = require("./Lambda");
Object.defineProperty(exports, "Lambda", { enumerable: true, get: function () { return __importDefault(Lambda_1).default; } });
var ServiceWrapper_1 = require("./ServiceWrapper");
Object.defineProperty(exports, "ServiceWrapper", { enumerable: true, get: function () { return ServiceWrapper_1.ServiceWrapper; } });
var CostAndUsageReports_1 = require("./CostAndUsageReports");
Object.defineProperty(exports, "CostAndUsageReports", { enumerable: true, get: function () { return __importDefault(CostAndUsageReports_1).default; } });
var AWSRegions_1 = require("./AWSRegions");
Object.defineProperty(exports, "AWS_REGIONS", { enumerable: true, get: function () { return AWSRegions_1.AWS_REGIONS; } });
var Recommendations_1 = require("./Recommendations");
Object.defineProperty(exports, "RightsizingRecommendations", { enumerable: true, get: function () { return Recommendations_1.RightsizingRecommendations; } });
Object.defineProperty(exports, "ComputeOptimizerRecommendations", { enumerable: true, get: function () { return Recommendations_1.ComputeOptimizerRecommendations; } });
