// @grnsft/if-unofficial-plugins/build/lib/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.WattTimeGridEmissions = exports.TeadsCurve = exports.TeadsAWS = exports.Co2js = exports.CloudCarbonFootprint = exports.BoaviztaCpuOutput = exports.BoaviztaCloudOutput = exports.AzureImporter = void 0;
var azure_importer_1 = require("./azure-importer");
Object.defineProperty(exports, "AzureImporter", { enumerable: true, get: function () { return azure_importer_1.AzureImporter; } });
var boavizta_1 = require("./boavizta");
Object.defineProperty(exports, "BoaviztaCloudOutput", { enumerable: true, get: function () { return boavizta_1.BoaviztaCloudOutput; } });
Object.defineProperty(exports, "BoaviztaCpuOutput", { enumerable: true, get: function () { return boavizta_1.BoaviztaCpuOutput; } });
var ccf_1 = require("./ccf");
Object.defineProperty(exports, "CloudCarbonFootprint", { enumerable: true, get: function () { return ccf_1.CloudCarbonFootprint; } });
var co2js_1 = require("./co2js");
Object.defineProperty(exports, "Co2js", { enumerable: true, get: function () { return co2js_1.Co2js; } });
var teads_aws_1 = require("./teads-aws");
Object.defineProperty(exports, "TeadsAWS", { enumerable: true, get: function () { return teads_aws_1.TeadsAWS; } });
var teads_curve_1 = require("./teads-curve");
Object.defineProperty(exports, "TeadsCurve", { enumerable: true, get: function () { return teads_curve_1.TeadsCurve; } });
var watt_time_1 = require("./watt-time");
Object.defineProperty(exports, "WattTimeGridEmissions", { enumerable: true, get: function () { return watt_time_1.WattTimeGridEmissions; } });
