// @cloud-carbon-footprint/aws/dist/lib/Recommendations/Rightsizing/RightsizingTargetRecommendation.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const common_1 = require("@cloud-carbon-footprint/common");
const RightsizingRecommendation_1 = __importDefault(require("./RightsizingRecommendation"));
class RightsizingTargetRecommendation extends RightsizingRecommendation_1.default {
    constructor(rightsizingRecommendationData) {
        super(rightsizingRecommendationData);
        const targetInstance = rightsizingRecommendationData.ModifyRecommendationDetail.TargetInstances.pop();
        this.accountId = rightsizingRecommendationData.AccountId;
        this.type = rightsizingRecommendationData.RightsizingType;
        this.region = this.getMappedRegion(targetInstance.ResourceDetails.EC2ResourceDetails.Region);
        this.instanceName = 'Recommend new instance';
        this.instanceType =
            targetInstance.ResourceDetails.EC2ResourceDetails.InstanceType;
        this.vCpuHours = this.getVCpuHours(targetInstance.ResourceDetails.EC2ResourceDetails);
        this.costSavings = parseFloat(targetInstance.EstimatedMonthlySavings);
        this.usageAmount = (0, common_1.getHoursInMonth)();
    }
}
exports.default = RightsizingTargetRecommendation;
