// @grnsft/if-plugins/build/lib/regex/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Regex = void 0;
const zod_1 = require("zod");
const helpers_1 = require("../../util/helpers");
const errors_1 = require("../../util/errors");
const validations_1 = require("../../util/validations");
const { InputValidationError, ConfigValidationError } = errors_1.ERRORS;
const Regex = (globalConfig) => {
    const errorBuilder = (0, helpers_1.buildErrorMessage)(exports.Regex.name);
    const metadata = {
        kind: 'execute',
    };
    /**
     * Checks global config value are valid.
     */
    const validateGlobalConfig = () => {
        if (!globalConfig) {
            throw new ConfigValidationError(errorBuilder({ message: 'Configuration data is missing' }));
        }
        const schema = zod_1.z.object({
            parameter: zod_1.z.string().min(1),
            match: zod_1.z.string().min(1),
            output: zod_1.z.string(),
        });
        return (0, validations_1.validate)(schema, globalConfig);
    };
    /**
     * Checks for required fields in input.
     */
    const validateSingleInput = (input, parameter) => {
        if (!input[parameter]) {
            throw new InputValidationError(errorBuilder({
                message: `\`${parameter}\` is missing from the input`,
            }));
        }
        return input;
    };
    /**
     * Executes the regex of the given parameter.
     */
    const execute = async (inputs) => {
        const safeGlobalConfig = validateGlobalConfig();
        const { parameter: parameter, match, output } = safeGlobalConfig;
        return inputs.map(input => {
            const safeInput = Object.assign({}, input, validateSingleInput(input, parameter));
            return {
                ...input,
                [output]: extractMatching(safeInput, parameter, match),
            };
        });
    };
    /**
     * Extracts a substring from the given input parameter that matches the provided regular expression pattern.
     */
    const extractMatching = (input, parameter, match) => {
        if (!match.startsWith('/') ||
            (match.endsWith('/g') && match.lastIndexOf('/') === 0)) {
            match = '/' + match;
            if (!match.endsWith('/g') || !match.endsWith('/')) {
                match += '/';
            }
        }
        const regex = eval(match);
        const matchedItem = input[parameter].match(regex);
        if (!matchedItem || !matchedItem[0]) {
            throw new InputValidationError(errorBuilder({
                message: `\`${input[parameter]}\` does not match the ${match} regex expression`,
            }));
        }
        return matchedItem[0];
    };
    return {
        metadata,
        execute,
    };
};
exports.Regex = Regex;
