// @grnsft/if-plugins/build/lib/divide/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Divide = void 0;
const zod_1 = require("zod");
const helpers_1 = require("../../util/helpers");
const errors_1 = require("../../util/errors");
const validations_1 = require("../../util/validations");
const { InputValidationError, ConfigValidationError } = errors_1.ERRORS;
const Divide = (globalConfig) => {
    const errorBuilder = (0, helpers_1.buildErrorMessage)(exports.Divide.name);
    const metadata = {
        kind: 'execute',
    };
    /**
     * Checks global config value are valid.
     */
    const validateGlobalConfig = () => {
        if (!globalConfig) {
            throw new ConfigValidationError(errorBuilder({ message: 'Configuration data is missing' }));
        }
        const schema = zod_1.z.object({
            numerator: zod_1.z.string().min(1),
            denominator: zod_1.z.string().or(zod_1.z.number().gt(0)),
            output: zod_1.z.string(),
        });
        return (0, validations_1.validate)(schema, globalConfig);
    };
    /**
     * Checks for required fields in input.
     */
    const validateSingleInput = (input, numerator, denominator) => {
        const schema = zod_1.z
            .object({
            [numerator]: zod_1.z.number(),
            [denominator]: zod_1.z.number().optional(),
        })
            .refine(_data => {
            if (typeof denominator === 'string' && !input[denominator]) {
                throw new InputValidationError(errorBuilder({
                    message: `\`${denominator}\` is missing from the input`,
                }));
            }
            return true;
        });
        return (0, validations_1.validate)(schema, input);
    };
    /**
     * Calculate the division of each input parameter.
     */
    const execute = async (inputs) => {
        const safeGlobalConfig = validateGlobalConfig();
        const { numerator, denominator, output } = safeGlobalConfig;
        return inputs.map(input => {
            const safeInput = Object.assign({}, input, validateSingleInput(input, numerator, denominator));
            return {
                ...input,
                [output]: calculateDivide(safeInput, numerator, denominator),
            };
        });
    };
    /**
     * Calculates the division of the given parameter.
     */
    const calculateDivide = (input, numerator, denominator) => input[numerator] / (input[denominator] || denominator);
    return {
        metadata,
        execute,
    };
};
exports.Divide = Divide;
