// @grnsft/if/build/util/validations.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.validate = exports.validateManifest = exports.manifestSchema = void 0;
const zod_1 = require("zod");
const errors_1 = require("./errors");
const aggregation_1 = require("../types/aggregation");
const parameters_1 = require("../types/parameters");
const { ManifestValidationError, InputValidationError } = errors_1.ERRORS;
/**
 * Validation schema for manifests.
 */
exports.manifestSchema = zod_1.z.object({
    name: zod_1.z.string(),
    description: zod_1.z.string().optional().nullable(),
    tags: zod_1.z
        .object({
        kind: zod_1.z.string().optional().nullable(),
        complexity: zod_1.z.string().optional().nullable(),
        category: zod_1.z.string().optional().nullable(),
    })
        .optional()
        .nullable(),
    aggregation: zod_1.z
        .object({
        metrics: zod_1.z.array(zod_1.z.string()),
        type: zod_1.z.enum(aggregation_1.AGGREGATION_METHODS),
    })
        .optional()
        .nullable(),
    params: zod_1.z
        .array(zod_1.z.object({
        name: zod_1.z.string(),
        description: zod_1.z.string(),
        aggregation: zod_1.z.enum(parameters_1.AGGREGATION_TYPES),
        unit: zod_1.z.string(),
    }))
        .optional()
        .nullable(),
    initialize: zod_1.z.object({
        plugins: zod_1.z.record(zod_1.z.string(), zod_1.z.object({
            path: zod_1.z.string(),
            method: zod_1.z.string(),
            'global-config': zod_1.z.record(zod_1.z.string(), zod_1.z.any()).optional(),
        })),
        outputs: zod_1.z.array(zod_1.z.string()).optional(),
    }),
    tree: zod_1.z.record(zod_1.z.string(), zod_1.z.any()),
    'if-version': zod_1.z.string().optional().nullable(),
});
/**
 * Validates given `manifest` object to match pattern.
 */
const validateManifest = (manifest) => (0, exports.validate)(exports.manifestSchema, manifest, ManifestValidationError);
exports.validateManifest = validateManifest;
/**
 * Validates given `object` with given `schema`.
 */
const validate = (schema, object, errorConstructor = InputValidationError) => {
    const validationResult = schema.safeParse(object);
    if (!validationResult.success) {
        throw new errorConstructor(prettifyErrorMessage(validationResult.error.message));
    }
    return validationResult.data;
};
exports.validate = validate;
/**
 * Beautify error message from zod issue.
 */
const prettifyErrorMessage = (issues) => {
    const issuesArray = JSON.parse(issues);
    return issuesArray.map((issue) => {
        const { code, path, message } = issue;
        const flattenPath = path.map(part => typeof part === 'number' ? `[${part}]` : part);
        const fullPath = flattenPath.join('.');
        if (code === 'custom') {
            return `${message.toLowerCase()}. Error code: ${code}.`;
        }
        return `"${fullPath}" parameter is ${message.toLowerCase()}. Error code: ${code}.`;
    });
};
