// @cloud-carbon-footprint/aws/dist/domain/AwsFootprintEstimationConstants.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.AWS_EMISSIONS_FACTORS_METRIC_TON_PER_KWH = exports.AWS_CLOUD_CONSTANTS = void 0;
const core_1 = require("@cloud-carbon-footprint/core");
const AWSRegions_1 = require("../lib/AWSRegions");
exports.AWS_CLOUD_CONSTANTS = {
    SSDCOEFFICIENT: 1.2,
    HDDCOEFFICIENT: 0.65,
    MEMORY_AVG: 80.69,
    MEMORY_BY_COMPUTE_PROCESSOR: {
        [core_1.COMPUTE_PROCESSOR_TYPES.CASCADE_LAKE]: 98.12,
        [core_1.COMPUTE_PROCESSOR_TYPES.SKYLAKE]: 81.32,
        [core_1.COMPUTE_PROCESSOR_TYPES.BROADWELL]: 69.65,
        [core_1.COMPUTE_PROCESSOR_TYPES.HASWELL]: 27.71,
        [core_1.COMPUTE_PROCESSOR_TYPES.COFFEE_LAKE]: 19.56,
        [core_1.COMPUTE_PROCESSOR_TYPES.SANDY_BRIDGE]: 16.7,
        [core_1.COMPUTE_PROCESSOR_TYPES.IVY_BRIDGE]: 9.67,
        [core_1.COMPUTE_PROCESSOR_TYPES.AMD_EPYC_1ST_GEN]: 89.6,
        [core_1.COMPUTE_PROCESSOR_TYPES.AMD_EPYC_2ND_GEN]: 129.78,
        [core_1.COMPUTE_PROCESSOR_TYPES.AWS_GRAVITON_2]: 129.78,
    },
    getMemory: (computeProcessors) => {
        const memoryForProcessors = computeProcessors.map((processor) => {
            return exports.AWS_CLOUD_CONSTANTS.MEMORY_BY_COMPUTE_PROCESSOR[processor];
        });
        const averageMemoryForProcessors = (0, core_1.getAverage)(memoryForProcessors);
        return averageMemoryForProcessors
            ? averageMemoryForProcessors
            : exports.AWS_CLOUD_CONSTANTS.MEMORY_AVG;
    },
    MIN_WATTS_AVG: 0.74,
    MIN_WATTS_BY_COMPUTE_PROCESSOR: {
        [core_1.COMPUTE_PROCESSOR_TYPES.CASCADE_LAKE]: 0.64,
        [core_1.COMPUTE_PROCESSOR_TYPES.SKYLAKE]: 0.65,
        [core_1.COMPUTE_PROCESSOR_TYPES.BROADWELL]: 0.71,
        [core_1.COMPUTE_PROCESSOR_TYPES.HASWELL]: 1,
        [core_1.COMPUTE_PROCESSOR_TYPES.COFFEE_LAKE]: 1.14,
        [core_1.COMPUTE_PROCESSOR_TYPES.SANDY_BRIDGE]: 2.17,
        [core_1.COMPUTE_PROCESSOR_TYPES.IVY_BRIDGE]: 3.04,
        [core_1.COMPUTE_PROCESSOR_TYPES.AMD_EPYC_1ST_GEN]: 0.82,
        [core_1.COMPUTE_PROCESSOR_TYPES.AMD_EPYC_2ND_GEN]: 0.47,
        [core_1.COMPUTE_PROCESSOR_TYPES.AWS_GRAVITON_2]: 0.47,
        [core_1.COMPUTE_PROCESSOR_TYPES.NVIDIA_K520]: 26,
        [core_1.COMPUTE_PROCESSOR_TYPES.NVIDIA_A10G]: 18,
        [core_1.COMPUTE_PROCESSOR_TYPES.NVIDIA_T4]: 8,
        [core_1.COMPUTE_PROCESSOR_TYPES.NVIDIA_TESLA_M60]: 35,
        [core_1.COMPUTE_PROCESSOR_TYPES.NVIDIA_TESLA_K80]: 35,
        [core_1.COMPUTE_PROCESSOR_TYPES.NVIDIA_TESLA_V100]: 35,
        [core_1.COMPUTE_PROCESSOR_TYPES.NVIDIA_TESLA_A100]: 46,
        [core_1.COMPUTE_PROCESSOR_TYPES.NVIDIA_TESLA_P4]: 9,
        [core_1.COMPUTE_PROCESSOR_TYPES.NVIDIA_TESLA_P100]: 36,
        [core_1.COMPUTE_PROCESSOR_TYPES.AMD_RADEON_PRO_V520]: 26,
    },
    getMinWatts: (computeProcessors) => {
        const minWattsForProcessors = computeProcessors.map((processor) => {
            return exports.AWS_CLOUD_CONSTANTS.MIN_WATTS_BY_COMPUTE_PROCESSOR[processor];
        });
        const averageWattsForProcessors = (0, core_1.getWattsByAverageOrMedian)(computeProcessors, minWattsForProcessors);
        return averageWattsForProcessors
            ? averageWattsForProcessors
            : exports.AWS_CLOUD_CONSTANTS.MIN_WATTS_AVG;
    },
    MAX_WATTS_AVG: 3.5,
    MAX_WATTS_BY_COMPUTE_PROCESSOR: {
        [core_1.COMPUTE_PROCESSOR_TYPES.CASCADE_LAKE]: 3.97,
        [core_1.COMPUTE_PROCESSOR_TYPES.SKYLAKE]: 4.26,
        [core_1.COMPUTE_PROCESSOR_TYPES.BROADWELL]: 3.69,
        [core_1.COMPUTE_PROCESSOR_TYPES.HASWELL]: 4.74,
        [core_1.COMPUTE_PROCESSOR_TYPES.COFFEE_LAKE]: 5.42,
        [core_1.COMPUTE_PROCESSOR_TYPES.SANDY_BRIDGE]: 8.58,
        [core_1.COMPUTE_PROCESSOR_TYPES.IVY_BRIDGE]: 8.25,
        [core_1.COMPUTE_PROCESSOR_TYPES.AMD_EPYC_1ST_GEN]: 2.55,
        [core_1.COMPUTE_PROCESSOR_TYPES.AMD_EPYC_2ND_GEN]: 1.69,
        [core_1.COMPUTE_PROCESSOR_TYPES.AWS_GRAVITON_2]: 1.69,
        [core_1.COMPUTE_PROCESSOR_TYPES.NVIDIA_K520]: 229,
        [core_1.COMPUTE_PROCESSOR_TYPES.NVIDIA_A10G]: 153,
        [core_1.COMPUTE_PROCESSOR_TYPES.NVIDIA_T4]: 71,
        [core_1.COMPUTE_PROCESSOR_TYPES.NVIDIA_TESLA_M60]: 306,
        [core_1.COMPUTE_PROCESSOR_TYPES.NVIDIA_TESLA_K80]: 306,
        [core_1.COMPUTE_PROCESSOR_TYPES.NVIDIA_TESLA_V100]: 306,
        [core_1.COMPUTE_PROCESSOR_TYPES.NVIDIA_TESLA_A100]: 407,
        [core_1.COMPUTE_PROCESSOR_TYPES.NVIDIA_TESLA_P4]: 76.5,
        [core_1.COMPUTE_PROCESSOR_TYPES.NVIDIA_TESLA_P100]: 306,
        [core_1.COMPUTE_PROCESSOR_TYPES.AMD_RADEON_PRO_V520]: 229,
    },
    getMaxWatts: (computeProcessors) => {
        const maxWattsForProcessors = computeProcessors.map((processor) => {
            return exports.AWS_CLOUD_CONSTANTS.MAX_WATTS_BY_COMPUTE_PROCESSOR[processor];
        });
        const averageWattsForProcessors = (0, core_1.getWattsByAverageOrMedian)(computeProcessors, maxWattsForProcessors);
        return averageWattsForProcessors
            ? averageWattsForProcessors
            : exports.AWS_CLOUD_CONSTANTS.MAX_WATTS_AVG;
    },
    NETWORKING_COEFFICIENT: 0.001,
    MEMORY_COEFFICIENT: 0.000392,
    PUE_AVG: 1.135,
    getPUE: () => {
        return exports.AWS_CLOUD_CONSTANTS.PUE_AVG;
    },
    AVG_CPU_UTILIZATION_2020: 50,
    REPLICATION_FACTORS: {
        S3: 6,
        S3_ONE_ZONE_REDUCED_REDUNDANCY: 2,
        EC2_EBS_VOLUME: 2,
        EC2_EBS_SNAPSHOT: 3,
        EFS: 3,
        EFS_ONE_ZONE: 2,
        RDS_BACKUP: 3,
        RDS_AURORA: 6,
        RDS_MULTI_AZ: 2,
        DOCUMENT_DB_BACKUP: 3,
        DOCUMENT_DB_STORAGE: 2,
        DYNAMO_DB: 2,
        ECR_STORAGE: 3,
        DOCUMENT_ELASTICACHE_BACKUP: 3,
        SIMPLE_DB: 2,
        DEFAULT: 1,
    },
    KILOWATT_HOURS_BY_SERVICE_AND_USAGE_UNIT: {
        total: {},
    },
    ESTIMATE_UNKNOWN_USAGE_BY: core_1.EstimateUnknownUsageBy.COST,
    SERVER_EXPECTED_LIFESPAN: 35040,
};
exports.AWS_EMISSIONS_FACTORS_METRIC_TON_PER_KWH = {
    [AWSRegions_1.AWS_REGIONS.US_EAST_1]: core_1.US_NERC_REGIONS_EMISSIONS_FACTORS.SERC,
    [AWSRegions_1.AWS_REGIONS.US_EAST_2]: core_1.US_NERC_REGIONS_EMISSIONS_FACTORS.RFC,
    [AWSRegions_1.AWS_REGIONS.US_WEST_1]: core_1.US_NERC_REGIONS_EMISSIONS_FACTORS.WECC,
    [AWSRegions_1.AWS_REGIONS.US_WEST_2]: core_1.US_NERC_REGIONS_EMISSIONS_FACTORS.WECC,
    [AWSRegions_1.AWS_REGIONS.US_GOV_EAST_1]: core_1.US_NERC_REGIONS_EMISSIONS_FACTORS.SERC,
    [AWSRegions_1.AWS_REGIONS.US_GOV_WEST_1]: core_1.US_NERC_REGIONS_EMISSIONS_FACTORS.WECC,
    [AWSRegions_1.AWS_REGIONS.AF_SOUTH_1]: 0.0009006,
    [AWSRegions_1.AWS_REGIONS.AP_EAST_1]: 0.00071,
    [AWSRegions_1.AWS_REGIONS.AP_SOUTH_1]: 0.0007082,
    [AWSRegions_1.AWS_REGIONS.AP_NORTHEAST_3]: 0.0004658,
    [AWSRegions_1.AWS_REGIONS.AP_NORTHEAST_2]: 0.0004156,
    [AWSRegions_1.AWS_REGIONS.AP_SOUTHEAST_1]: 0.000408,
    [AWSRegions_1.AWS_REGIONS.AP_SOUTHEAST_2]: 0.00076,
    [AWSRegions_1.AWS_REGIONS.AP_SOUTHEAST_3]: 0.0007177,
    [AWSRegions_1.AWS_REGIONS.AP_NORTHEAST_1]: 0.0004658,
    [AWSRegions_1.AWS_REGIONS.CA_CENTRAL_1]: 0.00012,
    [AWSRegions_1.AWS_REGIONS.CN_NORTH_1]: 0.0005374,
    [AWSRegions_1.AWS_REGIONS.CN_NORTHWEST_1]: 0.0005374,
    [AWSRegions_1.AWS_REGIONS.EU_CENTRAL_1]: 0.000311,
    [AWSRegions_1.AWS_REGIONS.EU_WEST_1]: 0.0002786,
    [AWSRegions_1.AWS_REGIONS.EU_WEST_2]: 0.000225,
    [AWSRegions_1.AWS_REGIONS.EU_SOUTH_1]: 0.0002134,
    [AWSRegions_1.AWS_REGIONS.EU_WEST_3]: 0.0000511,
    [AWSRegions_1.AWS_REGIONS.EU_NORTH_1]: 0.0000088,
    [AWSRegions_1.AWS_REGIONS.ME_SOUTH_1]: 0.0005059,
    [AWSRegions_1.AWS_REGIONS.ME_CENTRAL_1]: 0.0004041,
    [AWSRegions_1.AWS_REGIONS.SA_EAST_1]: 0.0000617,
    [AWSRegions_1.AWS_REGIONS.UNKNOWN]: 0.00039278188,
};
